### PR TITLE
CDAP-13354 fix incorrect response code for putting a profile

### PR DESCRIPTION
**Pull request by yaojiefeng**
JIRA: https://issues.cask.co/browse/CDAP-13354
Build: https://builds.cask.co/browse/CDAP-DUT6464-1

If we provide a json file like this: 
```{
  "provisioner": {
    "name": "gce-dataproc",
    "properties": [
      {
        "name": "accountKey",
        "value": ""
      },
      {
        "name": "region",
        "value": "global"
      },
      {
        "name": "zone",
        "value": "us-central1-a"
      },
      {
        "name": "projectId",
        "value": "pid37638"
      },
    ]
  }
} 
```
Gson will deserialize the list comma as a null property value, which will throw a NPE, we should throw it with a correct response code and message

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProfileHttpHandler.java [hunk 1]
```diff
@@ -182,6 +182,10 @@ private void validateProvisionerProperties(ProfileCreateRequest request) throws
     Collection<ProvisionerPropertyValue> provisionerProperties = provisionerInfo.getProperties();
     if (provisionerProperties != null) {
       for (ProvisionerPropertyValue value : provisionerProperties) {
+        if (value == null) {
+          throw new BadRequestException("A provisioner property cannot be null. Please make sure it is " +
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProfileHttpHandler.java]: I think it's better just to ignore it if null instead of making it an error. If I saw this error message, I would still be confused about what was wrong.

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProfileHttpHandler.java]: 👍

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProfileHttpHandlerTest.java [hunk 2]
```diff
@@ -109,6 +112,12 @@ public void testPutAndDeleteProfiles() throws Exception {
     ProvisionerDetail test = new ProvisionerDetail(spec.getName(), spec.getLabel(), spec.getDescription(),
                                                    new ArrayList<>());
     putProfile(NamespaceId.DEFAULT, test.getName(), test, 400);
+
+    // provide a profile with null provsioner property
+    List<ProvisionerPropertyValue> nullLists = Collections.singletonList(null);
```
Comment (albertshau) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProfileHttpHandlerTest.java]: I think this scenario deserves its own test case. Should also test when the list has a null mixed in with other properties.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProfileHttpHandler.java [hunk 3]
```diff
@@ -124,7 +124,8 @@ public void writeProfile(
     Profile profile =
       new Profile(profileName, profileCreateRequest.getDescription(),
                   profileId.getNamespaceId().equals(NamespaceId.SYSTEM) ? EntityScope.SYSTEM : EntityScope.USER,
-                  profileCreateRequest.getProvisioner());
+                  new ProvisionerInfo(profileCreateRequest.getProvisioner().getName(),
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProfileHttpHandler.java]: I think it's better to do this in ProfileCreateRequest's getProvisioner() method, to ensure it always returns a ProvisionerInfo without null properties.

**Review by albertshau — COMMENTED** [review]
small comment

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProfileHttpHandler.java (hunk 3 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProfileHttpHandler.java]: fixed

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProfileHttpHandlerTest.java (hunk 2 above)
Comment (yaojiefeng) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ProfileHttpHandlerTest.java]: Created a new test case for it

File: cdap-proto/src/main/java/co/cask/cdap/proto/profile/ProfileCreateRequest.java [hunk 4]
```diff
@@ -33,7 +33,7 @@ public ProfileCreateRequest(String description, ProvisionerInfo provisioner) {
   }
 
   public ProvisionerInfo getProvisioner() {
-    return provisioner;
+    return new ProvisionerInfo(provisioner.getName(), provisioner.getProperties());
```
Comment (albertshau) [inline cdap-proto/src/main/java/co/cask/cdap/proto/profile/ProfileCreateRequest.java]: please add a comment about why this is required.

**Review by albertshau — APPROVED** [review]
one minor comment about adding a comment, otherwise LGTM. Can merge after fixing.

File: cdap-proto/src/main/java/co/cask/cdap/proto/profile/ProfileCreateRequest.java (hunk 4 above)
Comment (yaojiefeng) [reply cdap-proto/src/main/java/co/cask/cdap/proto/profile/ProfileCreateRequest.java]: added
